### PR TITLE
ARROW-11730: [C++] Add implicit convenience constructors for constructing Future from Status/Result

### PR DESCRIPTION
**Pull request by lidavidm**
This enables use of macros like ARROW_RETURN_NOT_OK and ARROW_ASSIGN_OR_RAISE.

### CODE REVIEW

**Comment by lidavidm** [comment]
CC @westonpace and @bkietz who suggested this

File: cpp/src/arrow/util/future.h [hunk 1]
```diff
@@ -493,6 +493,22 @@ class ARROW_MUST_USE_TYPE Future {
     });
   }
 
+  /// \brief Implicit constructor to create a future from a Result, enabling use
+  ///     of macros like ARROW_ASSIGN_OR_RAISE.
+  Future(Result<ValueType> res) : Future() {  // NOLINT(runtime/explicit)
```
Comment (bkietz) [inline cpp/src/arrow/util/future.h:498]: Some tools don't recognize the parenthesized form
```suggestion
  Future(Result<ValueType> res) : Future() {  // NOLINT runtime/explicit
```

File: cpp/src/arrow/util/future.h [hunk 2]
```diff
@@ -493,6 +493,22 @@ class ARROW_MUST_USE_TYPE Future {
...
+      impl_ = FutureImpl::MakeFinished(FutureState::SUCCESS);
+    } else {
+      impl_ = FutureImpl::MakeFinished(FutureState::FAILURE);
+    }
+    SetResult(std::move(res));
+  }
+
+  /// \brief Implicit constructor to create a future from a Status, enabling use
+  ///     of macros like ARROW_RETURN_NOT_OK.
+  Future(Status s)  // NOLINT(runtime/explicit)
```
Comment (bkietz) [inline cpp/src/arrow/util/future.h:509]:
```suggestion
  Future(Status s)  // NOLINT runtime/explicit
```

File: cpp/src/arrow/util/future.h [hunk 3]
```diff
@@ -493,6 +493,22 @@ class ARROW_MUST_USE_TYPE Future {
     });
   }
```
Comment (bkietz) [inline cpp/src/arrow/util/future.h:494]: For completeness, we could also support construction from ValueType:

```suggestion
  }
  
  /// \brief Implicit constructor to create a finished future from a value
  Future(ValueType val) : Future() {  // NOLINT runtime/explicit
    impl_ = FutureImpl::MakeFinished(FutureState::SUCCESS);
    SetResult(std::move(val));
  }
```

File: cpp/src/arrow/util/future_test.cc [hunk 4]
```diff
@@ -392,6 +392,30 @@ TEST(FutureSyncTest, GetStatusFuture) {
...
+    AssertFailed(fut);
+    ASSERT_RAISES(Invalid, fut.result());
+  }
+  {
+    auto fut = ([]() -> Future<MoveOnlyDataType> {
+      return arrow::Result<MoveOnlyDataType>(arrow::Status::Invalid("Invalid"));
+    })();
+    AssertFailed(fut);
+    ASSERT_RAISES(Invalid, fut.result());
+  }
```
Comment (bkietz) [inline cpp/src/arrow/util/future_test.cc:410]: Test case for the previous:
```suggestion
  }
  {
    auto fut = ([]() -> Future<MoveOnlyDataType> { return MoveOnlyDataType(42); })();
    AssertSuccessful(fut);
  }
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Thanks! A few comments:

File: cpp/src/arrow/util/future.h [hunk 5]
```diff
@@ -493,6 +493,28 @@ class ARROW_MUST_USE_TYPE Future {
...
 
+  /// \brief Implicit constructor to create a finished future from a value
+  Future(ValueType val) : Future() {  // NOLINT runtime/explicit
+    impl_ = FutureImpl::MakeFinished(FutureState::SUCCESS);
+    SetResult(std::move(val));
+  }
+
+  /// \brief Implicit constructor to create a future from a Result, enabling use
+  ///     of macros like ARROW_ASSIGN_OR_RAISE.
+  Future(Result<ValueType> res) : Future() {  // NOLINT runtime/explicit
```
Comment (westonpace) [inline cpp/src/arrow/util/future.h:504]: Should we get rid of the static MakeFinished then?

Comment (lidavidm) [reply cpp/src/arrow/util/future.h:504]: IMO, it's probably better to be explicit in code; I'd rather this be purely to enable those convenience macros.

**Review by bkietz — APPROVED** [review]
Appveyor failure is resolved in master: https://issues.apache.org/jira/browse/ARROW-11724

merging

**Pull request closed without merging**